File: Programmers/Backtracking/광물 캐기.cpp
#include <iostream>
#include <string>
#include <vector>
#include <climits>

using namespace std;
// 0 : 다이아몬드, 1 : 철, 2 : 돌
vector<vector<int>> arr = {{1, 1, 1}, {5, 1, 1}, {25, 5, 1}};
int answer = INT_MAX;
void backtracking(vector<int> picks, vector<string> minerals, int idx, int cnt) {
    if (idx >= minerals.size() || (picks[0] == 0 && picks[1] == 0 && picks[2] == 0)) { // 다 캤거나 광물이 없을 때 종료
        answer = min(answer, cnt);
        return;
    }

    for (int i = 0; i < 3; i++) {
        if (picks[i] > 0) {
            int cur = 0;
            picks[i]--;
            for (int j = idx; j < min(idx + 5, (int) minerals.size()); j++) { // 주의 !!
                if (minerals[j] == "diamond") cur += arr[i][0];
                else if (minerals[j] == "iron") cur += arr[i][1];
                else if (minerals[j] == "stone") cur += arr[i][2];

            }
            backtracking(picks, minerals, idx + 5, cnt + cur);
            picks[i]++;
        }

    }
}

int solution(vector<int> picks, vector<string> minerals) {
    backtracking(picks, minerals, 0, 0);
    return answer;
}

File: Programmers/Backtracking/광물 캐기.py
answer = float('inf')


def backtracking(picks, minerals, dic, arr, tired, idx):
    global answer

    if tired >= answer: return  # 현재 피로도가 이미 answer보다 크면 스킵

    if idx >= len(minerals) or sum(picks) == 0:
        answer = min(answer, tired)
        return

    for i in range(3):
        if picks[i] > 0:

            cur = 0
            end = min(idx + 5, len(minerals))  # 곡괭이 하나 선택하면 5개 연속으로 써야
            for j in range(idx, end):
                cur += arr[i][dic[minerals[j]]]

            picks[i] -= 1
            backtracking(picks, minerals, dic, arr, tired + cur, end)
            picks[i] += 1


def solution(picks, minerals):
    global answer
    dic = {"diamond": 0, "iron": 1, "stone": 2}
    arr = [[1, 1, 1], [5, 1, 1], [25, 5, 1]]
    backtracking(picks, minerals, dic, arr, 0, 0)
    return answer